INT-4225: Add MessageConverters infrastructure

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4225

* To allow flexible way to convert incoming messages for the target arguments in method invocation
add `ConfigurableCompositeMessageConverter` global bean under `ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME`
* Retrieve that bean for the `MessageHandlerMethodFactor` in the `MessagingMethodInvokerHelper` to enrich afterwards target `HandlerMethodArgumentResolver`s
* The `ContentTypeConversionTests` demonstrates conversion incoming JOSN `string` into the target POJO argument in the service method

artembilan commented:
If it's OK with you, I can go ahead with the Docs on the matter.

OTOH the main idea that it can be overridden by the target application.
And one of the tricks is to add `MarshallingMessageConverter` from the SI XML module. So, when SI-XML is in CP, we have one more converter.

At the same time SCSt would be able to override it its own way with all those Cloud-aware converters, e.g. `TupleJsonMessageConverter` as it is now there.

garyrussell commented:
Looks good, although we should probably clearly document that conversion can only occur when there is only one candidate method.

Perhaps we could support `@Primary` and, if we find no matching methods for the payload type we invoke one marked with that annotation?

artembilan commented:
I like the idea with the `@Primary`! 😄 

Your thoughts about my proposal for automatic `MarshallingMessageConverter`?

garyrussell commented:
Makes sense (marshalling converter).

artembilan committed:
*  Add `@Primary` support
* Add documentation about `@Primary` and `contentType` conversion

garyrussell commented:
I wonder if it's ok to overload Spring's `@Primary` - the javadocs are wrong for this use case; OTOH  I don't think we want a second annotation - too confusing.

I wonder why travis keeps failing on SFTP.

garyrussell committed:
Doc Polishing

garyrussell commented:
Doc polish here - https://github.com/garyrussell/spring-integration/commit/99e029a3adb7b9ae6fc28d3442b4d6f48c4fb547

artembilan commented:
>  the javadocs are wrong for this use case;

Well, not necessarily. We create a method invocation bean anyway, so that might be still fine in this case to expect `@Primary`  together with the `@ServiceActivator`.

artembilan committed:
Change `@Primary` to `@Default`

artembilan committed:
INT-4225: Add MessageConverters infrastructure

JIRA: https://jira.spring.io/browse/INT-4225

* To allow flexible way to convert incoming messages for the target arguments in method invocation
add `ConfigurableCompositeMessageConverter` global bean under `ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME`
* Retrieve that bean for the `MessageHandlerMethodFactor` in the `MessagingMethodInvokerHelper` to enrich afterwards target `HandlerMethodArgumentResolver`s
* The `ContentTypeConversionTests` demonstrates conversion incoming JOSN `string` into the target POJO argument in the service method

*  Add `@Primary` support
* Add documentation about `@Primary` and `contentType` conversion

INT-1800: Add MTOM Support for WS

JIRA: https://jira.spring.io/browse/INT-1800

The Simple WebService Inbound and Outbound Gateways can operate with `WebServiceMessage`s directly.
This allows to create those messages manually and add attachments to them.
Or, on the other hand, process incoming messages with attachments manually.

For this purpose the `SimpleWebServiceOutboundGateway` is supplied with new `extractPayload` property.
Also the `UnmarshallingTransformer` can now process `MimeMessage` as payload to unmarshal it into object graph with attachments if that

INT-4225: Add MessageConverters infrastructure

JIRA: https://jira.spring.io/browse/INT-4225

* To allow flexible way to convert incoming messages for the target arguments in method invocation
add `ConfigurableCompositeMessageConverter` global bean under `ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME`
* Retrieve that bean for the `MessageHandlerMethodFactor` in the `MessagingMethodInvokerHelper` to enrich afterwards target `HandlerMethodArgumentResolver`s
* The `ContentTypeConversionTests` demonstrates conversion incoming JOSN `string` into the target POJO argument in the service method

artembilan committed:
Fix static imports for `@Default`
Add JavaDoc for the `SimpleWebServiceOutboundGateway.setExtractPayload()`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/util/MessagingMethodInvokerHelper.java:
```diff
@@ -545,6 +564,11 @@ else if (!Modifier.isPublic(method1.getModifiers())) {
 				}
 				return;
 			}
+			if (AnnotationUtils.getAnnotation(method1, Default.class) != null) {
```
Should we use `findAnnotation()` here, or do you think one level of meta-annotation is enough for this purpose?

I am leaning that way - having `@Default` on some obscure meta-meta-meta... could be confusing.

artembilan replied:
Well, my idea to always have it explicitly. If you think the method must be default, so make it like that with declaring `@Default` exactly on place!

garyrussell replied:
Agreed, but get __will__ handle one level of meta; merging...

garyrussell reviewed (commented):
Just that one comment/question and a couple of minor doc polishes which I'll do during merge after you answer.

garyrussell commented:
Merged as 28e8f23fd0ac04b538aeacdbccd3fbbf90517e91
